src/hooks/Recoil_Hooks.js

```diff
@@ -285,42 +286,57 @@ function useRecoilValueLoadable_MUTABLESOURCE<T>(
   }
   const storeRef = useStoreRef();
 
-  const getValue = useCallback(() => {
+  const getLoadable = useCallback(() => {
+    const store = storeRef.current;
+    const treeState = store.getState().currentTree;
+    return getRecoilValueAsLoadable(store, recoilValue, treeState);
+  }, [storeRef, recoilValue]);
+  const getLoadableWithTesting = useCallback(() => {
     if (__DEV__) {
       recoilComponentGetRecoilValueCount_FOR_TESTING.current++;
     }
-    return getRecoilValueAsLoadable(
-      storeRef.current,
-      recoilValue,
-      storeRef.current.getState().currentTree,
-    );
-  }, [storeRef, recoilValue]);
+    return getLoadable();
+  }, [getLoadable]);
 
   const componentName = useComponentName();
 
   const subscribe = useCallback(
-    (_something, callback) => {
+    (_storeState, callback) => {
       const store = storeRef.current;
       const subscription = subscribeToRecoilValue(
         store,
         recoilValue,
-        () => {
+        () =>
           Tracing.trace(
             'RecoilValue subscription fired',
             recoilValue.key,
             () => {
-              callback();
+              if (!gkx('recoil_suppress_rerender_in_callback')) {
+                return callback();
+              }
+              // Only re-render if the value has changed.
+              // This will evaluate the atom/selector now as well as when the
+              // component renders, but that may help with prefetching.
+              const newLoadable = getLoadable();
+              if (!prevLoadableRef.current.is(newLoadable)) {
+                callback();
+              }
             },
-          );
-        },
+          ),
         componentName,
       );
       return subscription.release;
     },
-    [recoilValue, storeRef, componentName],
+    [storeRef, recoilValue, componentName, getLoadable],
   );
 
-  return useMutableSource(useRecoilMutableSource(), getValue, subscribe);
+  const source = useRecoilMutableSource();
+  const loadable = useMutableSource(source, getLoadableWithTesting, subscribe);
+  const prevLoadableRef = useRef(loadable);
+  useEffect(() => {
```

Comment:
@drarmstr why not `layoutEffect`? Postponed update of the ref could cause some propblem, don't you think?